### PR TITLE
Ste flor/feature/138 clean

### PR DESCRIPTION
**Pull request by SteFlor**

### CODE REVIEW

**Comment by Unaimend** [comment]
If possible i'd also like to see test for each single function. I think you test all of them, but it's one huge test case. It's not easy to figure ou what is being tested. It would be nice if the test cases names correspond to what is tested?

**Comment by Unaimend** [comment]
To review the actual logic it would be nice to make an appointment so we can go through it together @SteFlor

**Comment by SteFlor** [comment]
We could talk about it today during or right after the main meeting @Unaimend

**Comment by Unaimend** [comment]
> We could talk about it today during or right after the main meeting @Unaimend

Today sounds fine

File: src/ModelMerger.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                if len(ex.metabolites.keys()) == 1:
+                    met = list(ex.metabolites.keys())[0]
+                    # create the metabolite in the translation compartment and give it a +1 stoichiometry (production)
+                    met_t = met.copy()
+                    #  TODO: replace regular expression with function call for the remove suffix function
+                    #   (need to modify the handle_metabolites_prefix_suffix_function)
+                    met_t.id = re.sub(r"[^_]+$", "t", met.id)  # substitutes the compartment suffix (all
+                                                                            # that follows the last underscore) with
+                                                                            # the new compartment's symbol
+                    met_t.compartment = "t"
```
Comment (SteFlor) [inline src/ModelMerger.py:37]: @SteFlor verify if line 37 automatically creates the new compartment, and describe this in a comment

File: src/ModelMerger.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        for met in to_translate:
+            met.id = re.sub(r"_t$", "", met.id)
+
+        sorted_matches = matches_df.sort_values(by=[score_type, 'source_namespace', 'target_namespace'],
+                                                ascending=[False, True, True])
+
+        best_matches = dict()
+
+        # Iterate over the to_translate list
+        for met_id in sorted_matches["source_namespace"]:
```
Comment (Unaimend) [inline src/ModelMerger.py:75]: Create a new with the property: if met_id in [m.id for m in to_translate]: This will make the code easier to read

File: src/ModelMerger.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            if met_id in [m.id for m in to_translate]:
+                #print([m.id for m in to_translate])
+                #print(met_id)
+                # Filter the matches for the current metabolite
+                matches_for_met = sorted_matches[sorted_matches['source_namespace'] == met_id]
+                # Assign the best available match
+                for _, row in matches_for_met.iterrows():
```
Comment (Unaimend) [inline src/ModelMerger.py:82]: By the sorting above matches_for_met will always contain the best match in the first row. So just pick the first row (Note: Is it possible that this df is empty?).

Comment (Unaimend) [reply src/ModelMerger.py:82]: Please refer in comment to the fact that this relies on the ordering established above

File: src/ModelMerger.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        reliable_matches = self.matches.loc[self.matches[score_type] >= score_thr]
+        #  selects the metabolites to be translated
+        to_translate = list()
+        for met in cobra_model.metabolites:
+            if met.compartment == "t":
+                to_translate = to_translate + [met]
+        # applies the translation to the metabolites
+        self.translate_metabolites(to_translate, reliable_matches, score_type)
+        #  translate the ids of the respective TR_ reactions
+        for met in cobra_model.metabolites:
```
Comment (Unaimend) [inline src/ModelMerger.py:124]: Please add a comment what this paragraph below is doing. Maybe with an exemplary metabolite?

File: src/ModelMerger.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                ex_rxn = cobra_model.reactions.get_by_id("EX_" + met.id)
+                tr_rxn = cobra_model.reactions.get_by_id("TR_" + met.id)
+                ex_rxn.lower_bound = tr_rxn.lower_bound
+                ex_rxn.upper_bound = tr_rxn.upper_bound
+                tr_rxn.lower_bound = -1000
+                tr_rxn.upper_bound = 1000
+
+
+    def translate_namespace(self):
+        self.convert_exchange_rxns_to_translation_rxns()
```
Comment (Unaimend) [inline src/ModelMerger.py:155]: Add an example what happens to a metabolite/reaction after each of those steps. Tell a little story of beautiful small metabolite on it journey through ModelMerging land

Comment (Unaimend) [reply src/ModelMerger.py:155]: I.e. ADd a comment after ther first function call that tells what happens to the ex reactions and why, and that we lose the actual ex reactions for now

File: src/ModelMerger.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            if met.compartment == "t":
+                tr_rxns_for_current_met = [rxn for rxn in met.reactions if rxn.id.startswith("TR_")]
+                if len(tr_rxns_for_current_met) != 1:
+                    raise ValueError(met.id + " should have one and only one TR_ reaction, it has " +
+                                     str(len(tr_rxns_for_current_met)) + " instead")
+                tr_rxn = tr_rxns_for_current_met[0]
+                tr_rxn.id = "TR_" + met.id
+
+    def create_exchanges(self):
+        cobra_model = self.memo_model.cobra_model
```
Comment (Unaimend) [inline src/ModelMerger.py:134]: Add a comment what this is doing and why (because we lost the actual EX_ in the "first" function

File: src/ModelMerger.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        for met in cobra_model.metabolites:
+            if met.compartment == "t":
+                reaction_id = "EX_" + met.id
+                rxn = cobra.Reaction(id=reaction_id, name=met.name + " exchange", lower_bound=-1000, upper_bound=1000)
+                rxn.add_metabolites({met: -1})
```
Comment (Unaimend) [inline src/ModelMerger.py:139]: why -1? What does this mean here? Which compartments are involved?

File: src/ModelMerger.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                cobra_model.add_reactions([rxn])
+
+    def set_rxn_bounds(self):
+        cobra_model = self.memo_model.cobra_model
```
Comment (Unaimend) [inline src/ModelMerger.py:143]: WHY are we doing this?

**Review by Unaimend — CHANGES_REQUESTED** [review]

**Comment by Unaimend** [comment]
Could you also add this step to the main function please

File: src/ModelMerger.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+                #print(met_id)
+                # Filter the matches for the current metabolite
+                matches_for_met = sorted_matches[sorted_matches['source_namespace'] == met_id]
+                # Assign the best available match
+                for _, row in matches_for_met.iterrows():
+                    match_id = row['target_namespace']
+                    if (met_id not in best_matches.keys()) and (match_id not in best_matches.values()):
+                        best_matches[met_id] = match_id
+                        #print(met_id + ": " + match_id + ", " + str(row[score_type]))
+                #else:
```
[the remark this answers is not in the document]
Comment (SteFlor) [reply src/ModelMerger.py:87]: It's commented because its function (to apply the best_matches translation when possible, and otherwise keep the original id for metabolites that don't have a translation) will be performed in a more readable way in lines 92 to 97

File: src/ModelMerger.py (hunk 3 above)
Comment (SteFlor) [reply src/ModelMerger.py:82]: > Please refer in comment to the fact that this relies on the ordering established above

Commented in commit "addresses https://github.com/Porthmeus/MeMoMe/pull/175/files/8656e8484be0fc3677bf76a13110c874253795a9#r1882320973", from line 116 to 118

**Comment by SteFlor** [comment]
> @SteFlor I am not super sure regarding the feasibility, but could add small tests. I'd like to see a few test that cover most edge cases with made up metabolites. At most 5 metabs per model

I would keep to be done as a separate issue, as it's not fully necessary to be covered now

**Comment by SteFlor** [comment]
> Could you also add this step to the main function please

What step?

**Comment by SteFlor** [comment]
> Could you also add this step to the main function please

Done in a commit

**Comment by Unaimend** [comment]
The merging step

File: tests/test_ModelMerger.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    def test_translate_namespace(self):
+        model1_path = self.dat.joinpath("dat/tiny_ecoli_keep_inchi.xml")
+        model2_path = self.dat.joinpath("manually_merged_models/gapseq_recon3D/M2_bacterial_model.xml")
+        model1 = MeMoModel.fromPath(model1_path)
+        model2 = MeMoModel.fromPath(model2_path)
+        matches = model1.match(model2)
+        merger = ModelMerger(model2, matches)
+        pre_translation_exchanges = model2.cobra_model.exchanges
+        merger.convert_exchange_rxns_to_translation_rxns()
+        self.assertTrue("t" in model2.cobra_model.compartments.keys())
```
[the remark this answers is not in the document]
Comment (SteFlor) [reply tests/test_ModelMerger.py:20]: As discussed today we postpone it for now

File: src/ModelMerger.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+                #    print(met_id + ": " + met_id)
+
+        # apply the id translation to the metabolite and re-appends the translation compartment suffix
+        for met in to_translate:
+            try:
+                met.id = best_matches[met.id] + "_t"
+            except KeyError:
+                met.id = met.id + "_t"
+
+    def translate_ids(self, score_thr, score_type="Name_score"):
```
Comment (Porthmeus) [inline src/ModelMerger.py:99]: default should be "total_score" not "Name_score"

Comment (Porthmeus) [reply src/ModelMerger.py:99]: We should remove the thresholding completely. We should not implement that here, but somewhere before. Here you can assume that you get a clean list of met1 to met2, which is also contains only unique matches. 

This should be a separate issue.
